phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
> Found another one that fails with both your updated version and with mine:

Good find!

> Here's another version:

> terrafrost@102d53b

> It's based on http://stackoverflow.com/a/9293146/569976

I added a few minor comments to it. But, more importantly, it also causes a "Fatal error:  Allowed memory size of ... bytes exhausted" if tested with the following on http://sandbox.onlinephpfunctions.com/ on the higher PHP versions, and doesn't return anything on the lower ones (note: the two `$z` lines need to be commented out for the lower versions):

```php
$a['key1'] = 'value1';
$a[] = &$a;

$b[] = &$a;
$b[] = &$b;

$c[] = &$b;

$z = [];
$z[] = [&$z];

echo phpseclib_safe_serialize($GLOBALS);
```

Here is a try to address all of that, and to also make the implementation smaller by bringing it closer to the above Stackoverflow example:
https://github.com/phpseclib/phpseclib/commit/047f2617f164f8f95d1c576994008719af68fecb

P.S.: This time just a commit as well (to stop the PR spam) - please let me know if you'd prefer a PR.